### PR TITLE
Use new bulk upload API for 'pages publish'

### PR DESCRIPTION
**Pull request by GregBrimble**
Another attempt at #940. There are a couple of tweaks made, but it's 90% the same. There's still some stuff to do behind-the-scenes, so keeping this PR as a draft for now.

Use new bulk upload API for 'wrangler pages publish', and only uploads files that are missing.

This raises the file limit back up to 20k for a deployment.

### CODE REVIEW

**Review by sidharthachatterjee — APPROVED** [review]
Just shipped the APIs and we're good to go now.

File: packages/wrangler/src/cfetch/internal.ts [hunk 1]
```diff
@@ -27,10 +27,11 @@ export async function fetchInternal<ResponseType>(
   resource: string,
   init: RequestInit = {},
   queryParams?: URLSearchParams,
-  abortSignal?: AbortSignal
+  abortSignal?: AbortSignal,
+  authOverride?: string
 ): Promise<ResponseType> {
   await requireLoggedIn();
-  const apiToken = requireApiToken();
+  const apiToken = authOverride || requireApiToken();
```
Comment (geelen) [inline packages/wrangler/src/cfetch/internal.ts:34]: Needed this because we can't pass an `init` with an `Authorization` header already: https://github.com/cloudflare/wrangler2/blob/d691a870e4131741f6480ee12c9024d0ccf0d2c0/packages/wrangler/src/cfetch/internal.ts#L104

If we relaxed that constraint we could back out a bunch of those changes. Thoughts?

Comment (geelen) [reply packages/wrangler/src/cfetch/internal.ts:34]: @petebacondarwin interested in your take on this...

File: package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "ioredis": "^4.28.2",
     "jest": "^27.5.1",
     "npm-run-all": "^4.1.5",
+    "p-queue": "^7.2.0",
```
Comment (petebacondarwin) [inline package.json:32]: This should be a `devDependency` in the `packages/wrangler/package.json` file.

The way to add deps to packages is from the root of the monorepo via

```sh
npm i -w wrangler -D p-queue
```

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 3]
```diff
@@ -32,7 +32,7 @@ const mocks: MockFetch<unknown>[] = [];
 export async function mockFetchInternal(
   resource: string,
   init: RequestInit = {},
-  queryParams: URLSearchParams = new URLSearchParams()
+  queryParams: URLSearchParams = new URLSearchParams(),
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:35]: Revert this change, prettier doesn't like it.

File: packages/wrangler/src/cfetch/internal.ts [hunk 4]
```diff
@@ -96,14 +96,13 @@ function requireApiToken(): string {
...
   apiToken: string
 ): void {
   if ("Authorization" in headers) {
-    throw new Error(
-      "The request already specifies an authorisation header - cannot add a new one."
-    );
+    // Only set if not already present
+    return;
   }
   headers["Authorization"] = `Bearer ${apiToken}`;
```
Comment (petebacondarwin) [inline packages/wrangler/src/cfetch/internal.ts:107]: NIT: slightly simpler code?

```
  if (!("Authorization" in headers) {
    headers["Authorization"] = `Bearer ${apiToken}`;
  }
```

File: packages/wrangler/src/pages.tsx [hunk 5]
```diff
@@ -284,7 +289,7 @@ function generateRulesMatcher<T>(
     T
   ][];
 
-  return ({ request }: { request: Request }) => {
+  return ({ request }: { request: MiniflareRequest }) => {
```
Comment (petebacondarwin) [inline packages/wrangler/src/pages.tsx:292]: Out of curiosity (OOC) - why the need to change these types?

File: packages/wrangler/src/pages.tsx [hunk 6]
```diff
@@ -1031,13 +1042,10 @@ const createDeployment: CommandModule<
...
-      content: Buffer;
-      metadata: Metadata;
-    };
-
-    type Metadata = {
-      sizeInBytes: number;
+    type FileContainer = {
+      content: string;
+      contentType: string;
+      size: number;
```
Comment (petebacondarwin) [inline packages/wrangler/src/pages.tsx:1048]: I quite liked having the unit of measurement easily visible (in the name `sizeInBytes`) since we have had problems before with things like time units being incorrect. If you don't want it in the name, then how about adding a JSDoc comment to this definition so it appears in the intellisense?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Looking good. Please can you fix the formatting and dependency issues.

File: package.json (hunk 2 above)
Comment (geelen) [reply package.json:32]: Should it? It's being used to manage concurrency on the Pages file uploads... It's not a dev dependency at all, I don't think?

File: packages/wrangler/src/pages.tsx (hunk 5 above)
Comment (geelen) [reply packages/wrangler/src/pages.tsx:292]: They're the same types as before, just renamed. I'm not sure as to why, @GregBrimble ?

Comment (petebacondarwin) [reply packages/wrangler/src/pages.tsx:292]: Oh right

**Review by petebacondarwin — APPROVED** [review]
LGTM!
